src/Adapter/Security/EmployeeAuthenticationHandler.php

```diff
@@ -0,0 +1,174 @@
+     */
+    public function __construct(
+        LegacyContext $legacyContext,
+        TokenStorageInterface $tokenStorage,
+        UserProviderInterface $userProvider,
+        CacheItemPoolInterface $cache,
+        RequestStack $requestStack
+    ) {
+        $this->legacyContext = $legacyContext;
+        $this->tokenStorage = $tokenStorage;
+        $this->userProvider = $userProvider;
+        $this->cache = $cache;
+        $this->requestStack = $requestStack;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function renewAuthenticationCredentials(EmployeeForAuthentication $employeeForAuthentication)
+    {
+        $email = $employeeForAuthentication->getEmail()->getValue();
+
+        $this->updateEmailInCookie($email);
+        $this->updatePasswordInCookie($employeeForAuthentication->getHashedPassword());
+        $this->updateSecurityToken($email);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setAuthenticationCredentials(EmployeeForAuthentication $employeeForAuthentication, $stayLoggedIn)
+    {
+        $employee = new Employee($employeeForAuthentication->getEmployeeId()->getValue());
+
+        // Assign logged in employee to the context
+        $this->legacyContext->getContext()->employee = $employee;
+
+        $request = $this->requestStack->getCurrentRequest();
+
+        if (null !== $request) {
+            $this->legacyContext->getContext()->employee->remote_addr = (int) ip2long($request->getClientIp());
+        }
+
+        // Save employee data to cookie
+        $cookie = $this->legacyContext->getContext()->cookie;
+        $cookie->id_employee = $employee->id;
+        $cookie->email = $employee->email;
+        $cookie->profile = $employee->id_profile;
+        $cookie->passwd = $employee->passwd;
+        $cookie->remote_addr = $employee->remote_addr;
+
+        if (!$stayLoggedIn) {
+            $cookie->last_activity = time();
+        }
+
+        $cookie->write();
+    }
+
+    /**
+     * Update employee password in cookie.
+     *
+     * @param string $hashedPassword
+     */
+    private function updatePasswordInCookie($hashedPassword)
+    {
+        $this->legacyContext->getContext()->cookie->passwd = $hashedPassword;
+        $this->legacyContext->getContext()->employee->passwd = $hashedPassword;
+        $this->legacyContext->getContext()->cookie->write();
+    }
+
+    /**
+     * Update employee email in cookie.
+     *
+     * @param string $email
+     */
+    private function updateEmailInCookie($email)
+    {
+        $this->legacyContext->getContext()->cookie->email = $email;
+        $this->legacyContext->getContext()->employee->email = $email;
+        $this->legacyContext->getContext()->cookie->write();
+    }
+
+    /**
+     * Update security token in token storage.
+     *
+     * @param string $email
+     */
+    private function updateSecurityToken($email)
+    {
+        $cacheKey = sprintf('app.employees_%s', sha1($email));
```

Comment:
> something bothers me loadUserByUsername is supposed to have username as a parameter, but here you use the email, and in some other places the username is used 

`loadUserByUsername()` is coming from `symfony/security` so I can't change this signature. 
I can change all places to be named `$username` instead of `$email`, but would that help? In PrestaShop email is the username, there is no such field as `username` in employee context. :thinking:

Reply:
I agree with @rokaszygmantas . Sadly the `loadUserByUsername` is indeed forced by Symfony and we better stick to it to make it easier to trace 😞